services/src/main/java/org/keycloak/services/DefaultComponentFactoryProviderFactory.java

```diff
@@ -0,0 +1,191 @@
+import org.jboss.logging.Logger;
+
+/**
+ * @author hmlnarik
+ */
+public class DefaultComponentFactoryProviderFactory implements ComponentFactoryProviderFactory {
+
+    private static final Logger LOG = Logger.getLogger(DefaultComponentFactoryProviderFactory.class);
+    public static final String PROVIDER_ID = "default";
+
+    private final AtomicReference<ConcurrentMap<String, ProviderFactory>> componentsMap = new AtomicReference<>(new ConcurrentHashMap<>());
+
+    /**
+     * Should an ID in the key be invalidated, it would invalidate also all the IDs in the values
+     */
+    private final ConcurrentMap<Object, Set<String>> dependentInvalidations = new ConcurrentHashMap<>();
+
+    private KeycloakSessionFactory factory;
+    private boolean componentCachingAvailable;
+
+    @Override
+    public void init(Scope config) {
+    }
+
+    @Override
+    public void postInit(KeycloakSessionFactory factory) {
+        this.factory = factory;
+        this.componentCachingAvailable = this.factory.getProviderFactory(ClusterProvider.class) != null;
+        if (! componentCachingAvailable) {
+            LOG.warn("No system-wide ClusterProviderFactory found. Cannot send messages across cluster, thus disabling caching of components.");
+        }
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public <T extends Provider> ProviderFactory<T> getProviderFactory(Class<T> clazz, String realmId, String componentId, Function<KeycloakSessionFactory, ComponentModel> modelGetter) {
+        ProviderFactory res = componentsMap.get().get(componentId);
+        if (res != null) {
+            LOG.tracef("Found cached ProviderFactory for %s in (%s, %s)", clazz, realmId, componentId);
+            return res;
+        }
+
+        // Apply the expensive operation before putting it into the cache
+        final ComponentModel cm;
+        if (modelGetter == null) {
+            LOG.debugf("Getting component configuration for component (%s, %s) from realm configuration", clazz, realmId, componentId);
+            AtomicReference<ComponentModel> cmr = new AtomicReference<>();
+            KeycloakModelUtils.runJobInTransaction(factory, session -> {
+                RealmModel realm = session.realms().getRealm(realmId);
+                cmr.set(realm == null ? null : realm.getComponent(componentId));
+            });
+            cm = cmr.get();
+        } else {
+            LOG.debugf("Getting component configuration for component (%s, %s) via provided method", realmId, componentId);
+            cm = modelGetter.apply(factory);
+        }
+
+        if (cm == null) {
+            return null;
+        }
+
+        final String provider = cm.getProviderId();
+        ProviderFactory<T> pf = provider == null
+          ? factory.getProviderFactory(clazz)
+          : factory.getProviderFactory(clazz, provider);
+
+        if (pf == null) {   // Either not found or not enabled
+            if (LOG.isDebugEnabled()) {
```

Comment:
Minor: Check for "debug", but log calls "log.warn" . It seems it should be either "isDebugEnabled/debug" or "isWarnEnabled/warn" ? Or check for "log.isDebugEnabled" removed entirely.